### PR TITLE
Adding max_retry to ActiveJob

### PR DESCRIPTION
**Pull request by cfiorini**
The possibility to change max attempts can be very useful if you want a job to die on first failure.
For now only Sidekiq and DelayedJob allow to set max attempts when adding a job to the queue.
max_retry accept a value or a block.

### CODE REVIEW

**Comment by jvanbaarsen** [comment]
Tested this locally, seems to work as advertised :+1: 
@rafaelfranca Can you add ActiveJob label?

File: activejob/lib/active_job/max_retry.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      # Specifies the max attempts for the job.
+      #
+      #   class PublishToFeedJob < ActiveJob::Base
+      #     max_retry 2
+      #
+      #     def perform(post)
+      #       post.to_feed!
+      #     end
+      #   end
+      def max_retry(max_retries=nil, &block)
```
Comment (vipulnsward) [inline activejob/lib/active_job/max_retry.rb]: We can be consistent and use `max_retry` everywhere instead of mixing `max_attempts` and `max_retry`

Comment (cfiorini) [inline activejob/lib/active_job/max_retry.rb]: @vipulnsward do you mean only commented lines ?

Comment (cfiorini) [inline activejob/lib/active_job/max_retry.rb]: btw i did on comments

Comment (vipulnsward) [inline activejob/lib/active_job/max_retry.rb]: @cfiorini everywhere. Else the usage of max_attempts vs max_retry might get confusing.

**Comment by cfiorini** [comment]
@vipulnsward now there is only max_retry :)

**Comment by jvanbaarsen** [comment]
Do we maybe want to log a warning if you're using max_retry with an adapter that is not supporting it?

**Comment by cfiorini** [comment]
@jvanbaarsen good point. Let me know if to include a warning.

BTW i'm still checking others adapter to find a way to set max_retry, for example: resque can be made through Resque::Failure::Base and Backburner has max_job_retries

**Comment by jvanbaarsen** [comment]
@seuros / @cristianbica  What do you think about the warning?

**Comment by cristianbica** [comment]
Basically I'm against this PR:
- there's no support in the majority of the providers
- providers implementation is different (is there a callback where one can send an email notification if a job fails?)
- I don't see too many use cases for this except web crawling or (maybe) some API communication
- retry is a very dangerous feature if not applied properly (how many times you've send a newsletter multiple times because somewhere around 80-90% of the list the job failed for some bug? :) )

Anyway it's not my call to reject this PR so if this gets merged definitely we need a warning message if the adapter does not support retries and I would also add test to test the retry functionality (only for the adapter that supports retries)

**Comment by cfiorini** [comment]
@cristianbica i see in a different way :)

* according to activejob doc there are only 6 adapters that handle retries and as i text before i'm checking for Resque and Backburner so at the end will be 4 on 6 and we will have the majority but in the mean time we can add a warning.

* max_retry just take a value to pass to the adapter like queue_as does. We can continue to handle email notification using rescue_from

* API communication is a good example because is a very large topic: if i have a limit usage i don't want to run out(also email services). Also imagine that you run a service that make remote checks 
and you want to handle number of retries for premium customer.

* retry can't be dangerous more of a job getting old in the queue :)

* retry functionality it's handled from the adapter so the test should be made in the gem adapter and not in AJ, we have just to test that is passed in the right way as queue name.

I make another consideration inspired from ActiveRecord:
If a database has a good feature you want in AR even if another one doesn't have it, a quick example can be column types support in Postgresql (BTW i'm a postgres fan) that Mysql and Sqlite3 don't have.

retry is feature that not all adapters has but this doesn't mean that should not be in AJ.

**Comment by cristianbica** [comment]
Let's not compare AR with AJ. AJ was designed as a wrapper around queueing adapters while AR is a full-fledged ORM and it's the M from MVC :)
My main objection is still about the different implementation of the retry:
- some adapter have a global option
- some have per job (for some you pass the option when enqueueing while for other you define it on the job class which means global flag for us)
- for que you need to set some instance variables
- the retry interval is different for each adapter (some linear, some exponential). 
One of the main ideas of AJ was that you can easy switch between adapter and getting the same behaviour.
This PR adds behaviour for only two. Qu doesn't supports retries (it's wrong in the rdoc/guides). 
About the testing I think we should have some tests (maybe in the integration ones) although we "test that is passed in the right way" (BTW we test queue_name in the integration). Anyway this can be added later.

cc @senny @matthewd @dhh

**Comment by matthewd** [comment]
As a general principle, I lean more towards us providing useful operations & options, particularly where a good number of the adapters can support them, rather than strict lowest common denominator. (I thus don't see anything wrong with the AR comparison here.)

I don't have enough relevant knowledge to speak to the feasibility of providing a semi-uniform "max retry" behaviour across the subset of adapters that have such a concept, though.

**Comment by rafaelsales** [comment]
Well, active_job implements retries itself, which by default will retry forever, right?

**Comment by dhh** [comment]
Going off the original motivation: "very useful if you want a job to die on first failure". We have an easy way to deal with this. Simply add a `rescue_from` handler to catch the exceptions you want to swallow, and do not call retry. Then the job is dead on first failure.

I could even see us institutionalize that pattern with something like `discard_job_on_exception` => `rescue_from(Exception) { log_job_death }`. But maybe that just needs to be documented.

**Comment by rafaelsales** [comment]
I think in general the ideal is in the middle: retry some times but not retry forever. An error regarding to ServiceUnavailable really makes sense to be retried like 10 times, but probably not more. Ideally this kind of error should actually have a exponential-backoff-based delay between retries because in many cases a service online becomes available again after a couple hours.

Beyond what's implemented in this PR, I think a global max retries setting is also useful to tackle delayed mails without having to implement a worker class just for that.

Is it something that would be welcome if implemented in a PR?

**Comment by dhh** [comment]
I'd be happy to entertain a PR. But I think we're going to have to put the
behavior into AJ. I'm already sorta regretting enqueued_at being an adapter
specific implementation.

**Comment by rafaelsales** [comment]
That's exactly the reason I feel it makes sense to have this as part of AJ: the way it deal with retries is by re-enqueuing a job regardless the adapter implementation, so that's just something else we can have under control without being adapter-specific. I'll try that to see if I can get better at OSS. Thanks for you attention here

**Comment by md5** [comment]
One thing I'll add is that the current retry behavior is already inconsistent as far as I know. If you don't add a `rescue_from` and you're using Delayed Job as your `:queue_adapter`, any job that throws an exception will be retried according to DJ's [built-in logic](https://github.com/collectiveidea/delayed_job#gory-details):

> The default Worker.max_attempts is 25. After this, the job either deleted (default), or left in the database with "failed_at" set. With the default of 25 attempts, the last retry will be 20 days later, with the last interval being almost 100 hours.

My understanding is that not all queue adapters/backends will do this, so this means that a simple change of `:queue_adapter` could change the behavior of a user's job from running once and throwing and exception into one that runs repeatedly.

Perhaps this deserves to be filed as a bug against the DJ queue adapter. It seems like the behavior of whether an underlying job gets retried our not should not be dependent on the chosen `:queue_adapter`.

**Comment by cristianbica** [comment]
If we implement this in AJ it's a different story than using the retries mechanism from adapters :).  ... although this departs AJ from it's initial scope of being just a wrapper around queueing backends.
@md5 I think DJ executes the `ActiveJob::Base#perform` and `rescue_from` will catch the exception before DJ. 
@dhh how would you see implementing `enqueue_at` without the adapter specific implementation and still work across all adapters?

**Comment by md5** [comment]
@cristianbica You're correct that `rescue_from` is able to catch any errors/exceptions before DJ sees them.

My point was more that if you *don't* implement `rescue_from`, DJ will use its retry default behavior. This means that a job without any `rescue_from` handlers may act differently under DJ than under other queue adapters since those other adapters may not requeue such jobs by default.

**Comment by cristianbica** [comment]
@md5 if retries mechanism is fully implemented in DJ then we will have to disable any retry mechanism from the adapters. On DJ you can do that https://github.com/collectiveidea/delayed_job#custom-jobs

**Comment by md5** [comment]
@cristianbica Yep. I think the DJ adapter should probably be setting `max_attempts` to `1` and pushing the burden of retries onto the user (or to AJ itself if it adds a retry mechanism).

**Comment by cristianbica** [comment]
we should depend on something the adapter maintainers can change. On DJ
case we can just implement max_attempts
on ActiveJob::QueueAdapters::DelayedJobAdapter::JobWrapper

-- 
Cristian Bica

**Comment by cfiorini** [comment]
@rafaelsales @dhh when I needed to handle max attempts  from AJ I was thinking to do that being adapter free and without using `rescue_from` but after I thought that AJ is a wrapper and a nice way was to implement `max_retry` that accepts block. 

@md5 in this PR i set `max_attempts` to 3 for DJ but, in theory, the right way is to use `Delayed::Worker::DEFAULT_MAX_ATTEMPTS` to keep what a developer expect from the adapter that is using. BTW i did for Sidekiq

@cristianbica do you think that with this PR we can achieve your thought about maintainers changes?

**Comment by md5** [comment]
Thanks @cfiorini. I should have looked closer at this PR.

**Pull request closed without merging**